src/dataflow/source/kafka.rs

```diff
@@ -288,54 +369,160 @@ where
+    // in next_partition_ts
     match timestamp_histories.borrow_mut().get_mut(id) {
         None => {}
         Some(entries) => {
-            while let Some((ts, offset)) = entries.first() {
-                let next_ts = ts + 1;
-                if last_processed_offset == *offset {
-                    entries.remove(0);
-                    cap.downgrade(&next_ts);
-                } else {
-                    // Offset isn't at a timestamp boundary, we take no action
-                    break;
+            for (pid, entries) in entries {
+                // Obtain the last offset processed (or -1 if no messages have yet been processed)
+                let last_offset = match last_processed_offset.get(pid) {
+                    Some(offs) => *offs,
+                    None => 0,
+                };
+                // Check whether timestamps can be closed on this partition
+                while let Some((partition_count, ts, offset)) = entries.first() {
+                    if partition_count > current_partition_count {
+                        // A new partition has been added, we need to update the appropriate
+                        // entries before we continue. This will also update the last_processed_offset
+                        // and next_partition_ts datastructures
+                        let partitions = update_partition_list(
+                            consumer,
+                            topic,
+                            *partition_count,
+                            last_processed_offset,
+                            next_partition_ts,
+                            *last_closed_ts,
+                        );
+                        *current_partition_count = i32::try_from(partitions.len()).unwrap();
+                    }
+                    if last_offset == *offset {
+                        // We have now seen all messages corresponding to this timestamp for this
+                        // partition. We
+                        // can close the timestamp (on this partition) and remove the associated metadata
+                        next_partition_ts.insert(*pid, *ts);
+                        entries.remove(0);
+                        changed = true;
+                    } else {
+                        // Offset isn't at a timestamp boundary, we take no action
+                        break;
+                    }
                 }
             }
         }
     }
+    //  Next, we determine the maximum timestamp that is fully closed. This corresponds to the minimum
+    //  timestamp across partitions. This value is stored in next_partition_ts
+    for next_ts in next_partition_ts.values() {
+        if *next_ts < min {
+            min = *next_ts
+        }
+    }
+    // Downgrade capability to new minimum open timestamp (which corresponds to min + 1).
+    if changed {
+        cap.downgrade(&(&min + 1));
```

Comment:
Will this panic if we see a `Timestamp::max_value()`? We probably want to drop the capability in that case?